### ISSUE TITLE
support 'conf.d' liked configurations, to modifying configs by adding or removing config files rather than modify the content of one config file

### ISSUE THREAD

**Issue by tedli** [prometheus/prometheus#5519]
## Proposal
**Use case. Why is this important?**

A system may want to reuse a prometheus server, so when a new module is added to the system, it had to modify prometheus config to let prometheus collect metrics. And when a module is no longer needed, we should remove related configs from prometheus along with removing the module itself.

Many tools support 'conf.d' liked configurations (nginx, supervisord, etc [many config files in a directory]), If prometheus could also support this, the use case above could achieve by adding or removing a config file rather than modify one config file's content.

*“Nice to have” is not a good use case. :)*

Manipulate complex config in one file is error prone.

**Comment by mashuiping** [comment]
Same as @tedli . By the way, is it possible to set alert configs by web? Since prometheus do not depend any storage, i think it can save configs by consul kv, just set a json string in consul, and prometheus get configs from consul kv.

**Comment by aixeshunter** [comment]
Good idea :)

**Comment by simonpasquier** [comment]
This topic has already been discussed several times in the past:
https://github.com/prometheus/prometheus/issues/1623#issuecomment-218443497
https://github.com/prometheus/prometheus/issues/1648
https://github.com/prometheus/prometheus/issues/1972#issuecomment-245882235

In general, it has been decided that the configuration is only defined via static files. It is also assumed that some kind of configuration management system (CMS) is in place to generate those files.

> Manipulate complex config in one file is error prone.

This is exactly why a CMS is useful.

For example, there are several projects that already deal with pulling configuration from K/V stores:
https://github.com/hashicorp/consul-template
https://github.com/kelseyhightower/confd

**Comment by tedli** [comment]
@simonpasquier 
Thanks for reply. I see. I'm going to close this issue. And I will do more searching before creating a new issue.

**Comment by simonpasquier** [comment]
@tedli this is fine, thanks for your understanding.

**Comment by webchi** [comment]
This is EPIC FAIL guys. Including config is a common linux approach for a many years. There is including files for alerts and no including for scrape_config... WUT
```
rule_files:
  - alerts/*.yml

scrape_configs:
  - ¯\_(ツ)_/¯
```

**Comment by roidelapluie** [comment]
This has been implemented in Prometheus 2.43, with scrape_config_files.
